sdks/python/apache_beam/io/iobase.py

```diff
@@ -1236,6 +1236,74 @@ def try_claim(self, position):
     raise NotImplementedError
 
 
+class WatermarkEstimator(object):
+  """A WatermarkEstimator which is used for estimating output_watermark based on
+  the timestamp of output records or manual modifications.
+
+  The base class provides common APIs that are called by the framework, which
+  are also accessible inside a DoFn.process() body. One derived watermark
+  estimator should implement all framework-called APIs, meanwhile, new added
+  APIs of the derived watermark estimator will also be exposed to a DoFn body.
+
+  Multi-threading safety is guarded by ThreadsafeWatermarkEstimator.
```

Comment:
```suggestion
  Internal state must not be updated asynchronously.
```

Reply:
We don't require synchronization when implementing `WatermarkEstimator`, right?

Reply:
We don't want users starting threads and updating the internal state outside of our knowledge.